src/NUnitFramework/tests/Constraints/NUnitEqualityComparerTests.cs

```diff
@@ -306,6 +308,89 @@ public void IEnumeratorIsDisposed()
             Assert.True(comparer.AreEqual(enumeration, enumeration, ref tolerance));
             Assert.That(enumeration.EnumeratorsDisposed);
         }
+
+        [TestCaseSource(nameof(GetRecursiveContainsTestCases))]
+        public void SelfContainedItemFoundInCollection<T>(T x, ICollection y)
+        {
+            var equalityComparer = new NUnitEqualityComparer();
+            var tolerance = Tolerance.Default;
+            var equality = equalityComparer.AreEqual(x, y, ref tolerance);
+
+            Assert.IsFalse(equality);
+            Assert.Contains(x, y);
+            Assert.That(y, Contains.Item(x));
+            Assert.That(y, Does.Contain(x));
+        }
+
+        [TestCaseSource(nameof(GetRecursiveComparerTestCases))]
+        public void SelfContainedItemDoesntRecurseForever<T>(T x, ICollection y)
+        {
+            var equalityComparer = new NUnitEqualityComparer();
+            var tolerance = Tolerance.Default;
+            equalityComparer.ExternalComparers.Add(new DetectRecursionComparer(30));
+
+            Assert.DoesNotThrow(() => equalityComparer.AreEqual(x, y, ref tolerance));
+        }
+
+        public static IEnumerable<TestCaseData> GetRecursiveComparerTestCases()
+        {
+            // Separate from 'GetRecursiveContainsTestCases' until a stackoverflow issue in 
+            // 'MsgUtils.FormatValue()' can be fixed for the below cases
+            foreach (var testCase in GetRecursiveContainsTestCases())
```

Comment:
As noted in the PR description, a separate stack overflow issues exists in `MsgUtils.FormatValue` for dictionary types that I'd like to defer fixing for a follow-up PR